### PR TITLE
[Messenger] Improve the profiler panel

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 4.1 
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes 
| Fixed tickets | N/A   
| License       | MIT
| Doc PR        | N/A

This is an attempt to enhance the profiler panel a bit.

**with the following messages dispatched:**

```php
$queryBus->dispatch(Envelope::wrap(new GetGreetingsQuery('Hello you'))
    ->with(new JustAFriendOfMine())
    ->with(new AndHisPlusO̶n̶e̶Eleven())
);
$commandBus->dispatch(new SendGiftCommand());
$queryBus->dispatch(new GetGreetingsQuery('Exterminate!'));
```

## Before

screenshot 2018-05-12 a 13 57 57

🐛calls order are wrong here, fixed in this PR

### collapsed

screenshot 2018-05-12 a 13 18 35
screenshot 2018-05-12 a 13 26 44

📝 _When loading the page, all messages details are collapsed by default but the first one per tab._

### expanded

screenshot 2018-05-12 a 13 49 42

### live

mai-12-2018 13-55-40

### toolbar (with exceptions)

screenshot 2018-05-10 a 23 18 32

## Notes

- Table headers are clickable, so you can jump directly to the message class in the code
- Reversing headers/rows allows to have a wider space for dumps and allows to add more entries in the future. This is an issue we already have with the Validator panel (when there is both an invalid value as object and a constraint violation dumped) which I'd like to revamp soon.
- ~~I wonder if we should keep the dispatched messages in call order, or if we can segregate by bus (using tabs?).~~
- ~~we could add a left container listing messages classes only, allowing to show details of a single message dispatched on a right container (similar to what the Form panel does). I'll probably suggest the same for the Validator panel.~~

### CODE REVIEW

File: src/Symfony/Component/Messenger/DataCollector/MessengerDataCollector.php [hunk 1]
```diff
@@ -79,38 +80,25 @@ private function collectMessage(string $busName, array $tracedMessage)
...
-                'object' => $this->cloneVar($message),
+                'type' => new ClassStub(\get_class($message)),
+                'value' => $message,
             ),
         );
 
         if (array_key_exists('result', $tracedMessage)) {
             $result = $tracedMessage['result'];
-
-            if (\is_object($result)) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/DataCollector/MessengerDataCollector.php]: I don't feel like we need all these conditions and formats... let me know if there was a reason.

**Comment by sroze** [comment]
> I wonder if we should keep the dispatched messages in call order, or if we can segregate by bus (using tabs?).

👍 to keep it by call order. It makes much more sense when you want to investigate what happened (which is the point of the profiler)

**Comment by ogizanagi** [comment]
@sroze : Then what about a default "All" tab ? :)

**Comment by sroze** [comment]
> Reversing headers/rows allows to have a wider space for dumps and allows to add more entries in the future

Maybe we can have a collapsed/expanded version? The collapsed will just be the bus name and the message name. You can click on it and it displays this tabled version?

**Comment by ro0NL** [comment]
What about a filter and perhaps share some logic with https://github.com/symfony/symfony/pull/24263?

**Comment by ogizanagi** [comment]
New version pushed & screens updated in PR description.

@sroze : A simple collapsible version is indeed better. Long messages' FQCN wouldn't suit great for what I had in mind first, and the amount of messages per request would never be big, so it's a great idea!

@ro0NL : when #24263 let's see what we can do here. For now, tabs are perfect to me :)

File: src/Symfony/Component/Messenger/TraceableMessageBus.php [hunk 2]
```diff
@@ -16,6 +16,8 @@
  */
 class TraceableMessageBus implements MessageBusInterface
 {
+    private static $calls = 0;
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/TraceableMessageBus.php]: This might not be ideal. Use a microtime to keep calls order instead?

**Comment by yceruto** [comment]
This look great! just a minor comment, making bg-red each counter with exception is a bit confusing at first sight, is there other way to represent that the tab has exceptions?

**Comment by ogizanagi** [comment]
> making bg-red each counter with exception is a bit confusing at first sight, is there other way to represent that the tab has exceptions?

Not sure. Actually that's kind of what we're already used to when there is an "error". For instance any item in the toolbar or the menu.
The "Infos & Error tabs" in the logger panel only shows the nb of error logs instead of the whole count when there is an error, so looses an info. 
Not sure showing both total & with exceptions counts would render nicely neither.

**Comment by yceruto** [comment]
For example, could we show a `border-bottom: 2px solid #B0413E;` with grey bg if not all are exceptions and show as now if all of them are exceptions?

**Comment by yceruto** [comment]
Another sample of the same:
messages_color

File: src/Symfony/Component/Messenger/TraceableMessageBus.php (hunk 2 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/TraceableMessageBus.php]: What's wrong with keeping PHP incrementing the keys with `$this->dispatchedMessages[]`?

**Review by sroze — COMMENTED** [review]
That looks amazing ❤️

File: src/Symfony/Component/Messenger/TraceableMessageBus.php (hunk 2 above)
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/TraceableMessageBus.php]: There are multiple buses, so multiple traceable buses. This ensures the calls order is kept, while iterating over each traceable buses in the datacollector to get their dispatched messages separately won't 😉

**Comment by mvrhov** [comment]
IMO "All buses" could easily be shortened to just "All"

File: src/Symfony/Component/Messenger/TraceableMessageBus.php (hunk 2 above)
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/TraceableMessageBus.php]: That's what I suggested in https://github.com/symfony/symfony/pull/27202#discussion_r187465603 :p
I take that as a "Let's do it" then 😄 Thanks

Comment (sroze) [reply src/Symfony/Component/Messenger/TraceableMessageBus.php]: 🙊

**Comment by sroze** [comment]
The tests are failing now by the way :)

**Comment by ogizanagi** [comment]
@sroze: I know :) I'll try to make this ready this afternoon.

**Comment by sroze** [comment]
No worries, no pressure at all.

**Comment by yceruto** [comment]
@ogizanagi what about don't show the "Exception" row if "No Exception"?

**Comment by ogizanagi** [comment]
Done :)

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/messenger.html.twig [hunk 3]
```diff
@@ -24,49 +40,128 @@
...
+        </div>
+
+        {% for bus in collector.buses %}
+        <div class="tab message-bus">
+            {% set messages = collector.messages(bus) %}
+            {% set exceptionsCount = collector.exceptionsCount(bus) %}
+            <h3 class="tab-title">{{ bus }}<span class="badge {{ exceptionsCount ? exceptionsCount == messages|length ? 'status-error' : 'status-some-errors' }}">{{ messages|length }}</span></h3>
+
+            <div class="tab-content">
+                <p class="text-muted">Ordered list of messages dispatched by the <code>{{ bus }}</code> bus</p>
```
Comment (sroze) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/messenger.html.twig]: `...dispatch *on* the ...`?

**Review by sroze — APPROVED** [review]
Amazing work. Thank you!

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/messenger.html.twig (hunk 3 above)
Comment (ogizanagi) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/messenger.html.twig]: Fixed!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ogizanagi.

**Pull request merged**